platform-tests/src/test/java/org/junit/platform/commons/util/ClasspathScannerTests.java

```diff
@@ -166,14 +192,39 @@ private void scanForClassesInClasspathRootWithinJarFile(String resourceName) thr
 			var classpathScanner = new ClasspathScanner(() -> classLoader, ReflectionUtils::tryToLoadClass);
 
 			var classes = classpathScanner.scanForClassesInClasspathRoot(jarfile.toURI(), allClasses);
-			var classNames = classes.stream().map(Class::getName).collect(Collectors.toList());
-			assertThat(classNames).hasSize(3) //
-					.contains("org.junit.platform.jartest.notincluded.NotIncluded",
+			assertThat(classes).extracting(Class::getName) //
+					.containsExactlyInAnyOrder("org.junit.platform.jartest.notincluded.NotIncluded",
 						"org.junit.platform.jartest.included.recursive.RecursivelyIncluded",
 						"org.junit.platform.jartest.included.Included");
 		}
 	}
 
+	@Test
+	void scanForResourcesInClasspathRootWithinJarFile() throws Exception {
+		scanForResourcesInClasspathRootWithinJarFile("/jartest.jar");
+	}
+
+	@Test
+	void scanForResourcesInClasspathRootWithinJarWithSpacesInPath() throws Exception {
+		scanForResourcesInClasspathRootWithinJarFile("/folder with spaces/jar test with spaces.jar");
+	}
+
+	private void scanForResourcesInClasspathRootWithinJarFile(String resourceName) throws Exception {
+		var jarfile = getClass().getResource(resourceName);
+
+		try (var classLoader = new URLClassLoader(new URL[] { jarfile })) {
+			var classpathScanner = new ClasspathScanner(() -> classLoader, ReflectionUtils::tryToLoadClass);
+
+			var resources = classpathScanner.scanForResourcesInClasspathRoot(jarfile.toURI(), allResources);
+			assertThat(resources).extracting(Resource::getName) //
+					.containsExactlyInAnyOrder("org/junit/platform/jartest/notincluded/not-included.resource",
+						"org/junit/platform/jartest/included/included.resource",
+						"org/junit/platform/jartest/included/recursive/recursively-included.resource",
+						// TODO: This is interesting. Would we also scan classes in META-INF/versions?
+						"META-INF/MANIFEST.MF");
```

Comment:
I'm not sure if this is surfacing a bug in the class-path scanner or if this is expected behavior.

Reply:
Yes, looks like we would also find classes in `META-INF/versions` but then fail to load them in `ClasspathScanner.processClassFileSafely`. Let's ignore this for this PR, though.